### PR TITLE
feat(loss): add AMSE loss with SHT caching to address spectral double penalty

### PR DESCRIPTION
**Pull request by ananya12k**
# Pull Request

## Description

This PR implements a modified **Spherical Harmonic Loss Function** to address the double penalty problem in data-driven weather forecasting, as discussed in:

**[Paper] Fixing the Double Penalty in Data-Driven Weather Forecasting Through a Modified Spherical Harmonic Loss Function**  
Fixes: [#159](https://github.com/openclimatefix/graph_weather/issues/159)

### Summary of Changes:
- Implemented the Angular Mean Squared Error (AMSE) loss using Spherical Harmonic Transform (SHT).
- Added caching for spherical transforms to reduce redundant computation.
- Registered AMSE loss in the model training configuration for easy selection.
- Created a Colab notebook for testing basic loss behavior, shapes, and gradients.

## How Has This Been Tested?

- Verified AMSE output values using synthetic tensors.
- Checked gradient shapes via `.backward()` on dummy data.
- Compared output loss values across batch sizes and spatial resolutions.
- Tested in both local development environment and Google Colab.

All tests were run locally and in Google Colab.  
Notebook for verification: [Google Colab](https://colab.research.google.com/drive/1tkGc66OJ0N4JuAkl_xGKAxfsKlwA3fK8?usp=sharing)

- [x] Yes, initial behavior tested manually.

_If your changes affect data processing, have you plotted any changes?_

- [ ] No plots yet — basic functional validation only.

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (where applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (where applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by jacobbieker — CHANGES_REQUESTED** [review]
This looks really good! The one change I would say this needs would be to add unit tests for the loss in the repo. The colab is nice, but it would be better to have it run each time.

File: tests/test_asme_loss.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    return (torch.rand(num_channels) + 0.5).clone().detach()
+
+
+@pytest.fixture
+def loss_fn(feature_variance):
+    """Provides an instance of the AMSENormalizedLoss."""
+    return AMSENormalizedLoss(feature_variance=feature_variance)
+
+
+class TestAMSENormalizedLoss:
```
Comment (jacobbieker) [inline tests/test_asme_loss.py]: We use the pytest-style of tests, so please remove the test class, and just have the test functions. You can see the other test files for comparison.

Comment (ananya12k) [reply tests/test_asme_loss.py]: @jacobbieker I've updated the tests to follow the pytest style by removing the class and converting them into standalone test functions, as per the convention in the other test files. Let me know if there's anything else you'd like me to adjust.

File: tests/test_asme_loss.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
+import pytest
+import torch
+import torch_harmonics as th
+
+from graph_weather.models.losses import AMSENormalizedLoss
+
+# This file contains tests for the AMSENormalizedLoss class, which computes the
```
Comment (jacobbieker) [inline tests/test_asme_loss.py]:
```suggestion
```

**Review by jacobbieker — APPROVED** [review]
Looks good to me! Thanks for this, just remove that one comment, and I can then merge it!

File: tests/test_asme_loss.py (hunk 2 above)
Comment (ananya12k) [reply tests/test_asme_loss.py]: @jacobbieker Ok noted and resolved, I have removed that comment.
